google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -101,26 +105,33 @@ public int read(ByteBuffer dst) throws IOException {
       checkOpen();
       int amt;
       int retries = 0;
-      int maxRetries = 3;
+      int maxRetries = Math.max(3, maxChannelReopens);
       dst.mark();
       while (true) {
         try {
           dst.reset();
           amt = channel.read(dst);
           break;
         } catch (StorageException exs) {
-          if (exs.getMessage().contains("Connection closed prematurely") && reopens < maxChannelReopens) {
-            // this error isn't marked as retryable since the channel is closed;
-            // but here at this higher level we can retry it.
+          if (isReopenable(exs)) {
+            // these errors aren't marked as retryable since the channel is closed;
+            // but here at this higher level we can retry them.
             reopens++;
+            if (reopens > maxChannelReopens) {
+              throw new StorageException(exs.getCode(), "All reopens failed", exs);
+            }
             sleepForAttempt(reopens);
             innerOpen();
             continue;
-          } else if ((exs.getCode() == 500 || exs.getCode() == 503)  && retries < maxRetries) {
+          } else if (exs.isRetryable() || exs.getCode() == 500 || exs.getCode() == 503) {
             retries++;
+            if (retries > maxRetries) {
+              throw new StorageException(exs.getCode(), "All retries failed", exs);
```

Comment:
I'm not sure on the spelling either! I added a comment to document this potentially confusing case.